src/gather_classes.rs

```diff
@@ -257,80 +254,123 @@ fn update_report(
             continue;
         }
-        let calls = get_calls(root_call).context("Failed to get calls from root_call")?;
-        for call in calls {
-            let maybe_kvp = get_tuple_from_call(call);
-            match maybe_kvp {
-                Ok(kvp) => {
-                    report.add_fields(kvp.1, kvp.0);
-                    updates += 1;
-                }
-                Err(err) => {
-                    debug!("Failed to extract tuple with error: {err}");
-                }
-            }
+        let calls =
+            get_calls_with_count(root_call).context("Failed to get calls from root_call")?;
+        for ((entry_point, class_hash), count) in calls {
+            report.update_count(class_hash, entry_point, count);
+            updates += count;
         }
     }
     assert!(report.blocks.insert(block_num));
     Ok(updates)
 }
 
-// Converts recursive calls to value.get("calls") to a boxed iterator.
-fn get_calls<'a>(obj: &'a Value) -> Result<Box<dyn Iterator<Item = &Value> + 'a>, anyhow::Error> {
+/// Merges two HashMaps of calls into a single HashMap of calls. We only consider a call if
+/// it exists in both maps.
+///
+/// This function updates the `result` HashMap in-place with the merged calls.
+fn merge_calls_with_count(
+    base_calls: &HashMap<CallKey, usize>,
+    native_calls: &HashMap<CallKey, usize>,
+    mut result: HashMap<CallKey, usize>,
+) -> HashMap<CallKey, usize> {
+    for (call_key, base_count) in base_calls {
+        if native_calls.contains_key(call_key) {
+            result
+                .entry(*call_key)
+                .and_modify(|count| *count += base_count)
+                .or_insert(*base_count);
+        }
+    }
+    result
+}
+
+/// Recursively extracts calls with their counts from a [Value] object representing a transaction trace.
+///
+/// This traverses the object recursively and extracts calls with their counts. It also handles the case
+/// where the calls are Different.
+fn get_calls_with_count(obj: &Value) -> Result<HashMap<CallKey, usize>, anyhow::Error> {
     // TODO (#72) Put debug logging here under a core::option_env flag so it is normally hidden.
-    match obj {
-        Value::String(string) => {
-            if string_is_same(string) {
-                Ok(Box::new(::std::iter::empty()))
-            } else {
-                Err(anyhow!("unexpected string: {}", string))
+    fn get_calls_inner(
+        obj: &Value,
+        mut result: HashMap<CallKey, usize>,
+    ) -> Result<HashMap<CallKey, usize>, anyhow::Error> {
+        match obj {
+            Value::String(string) => {
+                // If the string is SAME or EMPTY, then there are no calls to extract and continue processing
+                if string_is_same(string) || string_is_empty(string) {
+                    Ok(result)
+                } else {
+                    Err(anyhow!("String is not SAME or EMPTY: {}", string))
+                }
```

Comment:
Sorry didn't notice before, but this case, after the previous change, becomes redundant. 

All none matching gases are handled in the branch `_ => { .. } `